doc/tools/docgram/common.edit_mlg

```diff
@@ -122,48 +82,176 @@ DELETE: [
 | test_ident_no_do
 | ssrdoarg  (* todo: this and the next one should be removed from the grammar? *)
 | ssrseqdir
+
+(* unused *)
+| constr_comma_sequence'
+| auto_using'
+| constr_may_eval
+]
+
+ssrintrosarg: [ | DELETENT ]
+
+(* additional nts to be spliced *)
+
+hyp: [
+| var
+]
+
+tactic_then_gen: [
+| EDIT ADD_OPT tactic_expr5 "|" tactic_then_gen
+| EDIT ADD_OPT tactic_expr5 ".." tactic_then_last
+| REPLACE OPT tactic_expr5 ".." tactic_then_last
+| WITH ltac_expr_opt ".." or_opt ltac_expr_opt_list_or
+]
+
+ltac_expr_opt_list_or: [
+| ltac_expr_opt_list_or "|" OPT tactic_expr5
+| OPT tactic_expr5
 ]
 
+reference: [ | DELETENT ]
+
+reference: [
+| qualid
+]
+
+fullyqualid: [ | DELETENT ]
+
+fullyqualid: [
+| qualid
+]
+
+basequalid: [
+| REPLACE ident fields
+| WITH qualid field
+]
+
+fields: [ | DELETENT ]
+
+dirpath: [
+| REPLACE ident LIST0 field
+| WITH ident
+| dirpath field
+]
+
+binders: [
+| DELETE Pcoq.Constr.binders  (* todo: not sure why there are 2 "binders:" *)
+]
+
+lconstr: [
+| DELETE l_constr
+]
+
+return_type: [
+| REPLACE OPT [ OPT [ "as" name ] case_type ]
+| WITH as_name_opt case_type
+| empty
+]
+
+as_name_opt: [
+| "as" name
+| empty
+]
+
+as_dirpath: [
+| "as" dirpath
+| empty
+]
+
+operconstr100: [
+| REPLACE operconstr99 "<:" binder_constr
+| WITH operconstr99 "<:" term
```

Comment:
What is the purpose of this replacement? If the goal is to replace `binder_constr` with `term` everywhere, can't you just rename `binder_constr` into `term`? Right now this creates an incomprehensible grammar. `binder_constr` is defined, used very little and very inconsistently.

Reply:
In later edits, `operconstr200` gets renamed to `term` (and the other `operconstr*` -> `term*`).
I confused things here by using the renamed nonterminal rather than the one that appears in fullGrammar.  Using `operconstr200` here instead of  `term` should be clearer, which I will do.

The convention is to do renames last so edits are easily related to fullGrammar.  Also, don't cheat by the renamed token.  Maybe I can change the tool to detect this.

Given this definition of `operconstr200`, replacing the two lines in operconstr99 with the single line `operconstr99 "<:" operconstr200` is clearly equivalent, right?

```
operconstr200: [
| binder_constr
| operconstr100
]

operconstr100: [
| operconstr99 "<:" binder_constr
| operconstr99 "<:" operconstr100
```

Reply:
I would think so. If that's actual code, would you mind opening a PR changing it? Let's try to use this opportunity to simplify whatever weirdness we find in the actual code.

Reply:
Actually, I am starting to suspect that there might be a bug in the way your tool translated the original grammar into `fullGrammar` which explains why you see such weird things. The original rule is:

https://github.com/coq/coq/blob/6ed3b02af77313d62ec868b4a88a208a9003857d/parsing/g_constr.mlg#L195-L202

And the tool translated it to:

https://github.com/coq/coq/blob/6ed3b02af77313d62ec868b4a88a208a9003857d/doc/tools/docgram/fullGrammar#L70-L77

But `SELF` does not necessarily represent the present entry at the same level. According to the [documentation of camlp5](https://camlp5.github.io/doc/htmlc/grammars.html):

>the identifier "SELF" which is a recursive call to the present entry, according to the associativity (i.e. it may be a call at the current level, to the next level, or to the top level of the entry): "SELF" is equivalent to the name of the entry itself

@ppedrot or @herbelin might be able to confirm this. I will ask Hugo in person if I can find him.

Reply:
I asked Hugo, it turns out that in this case at least your translation is correct, and the simplification we talked about could be applied directly in the `.mlg` file.

Reply:
Submitted as #10963